### DISCUSSION TITLE
API / Extension send msg

### DISCUSSION THREAD

**Discussion by AppDesv** [LiveHelperChat/livehelperchat#2043] [General]
Hello Remdex,

I would like to know if there is an available route in the LHC API for sending messages through a webhook, similar to the screenshot functionality.

![Captura de Tela 2024-01-26 às 10 19 27](https://github.com/LiveHelperChat/livehelperchat/assets/142036461/869f5d44-6cd9-4f1c-8083-261de54b629c)

If such a route does not exist, I am interested in whether it is possible to create one or develop an extension module that would allow standard LHC users to send messages through a webhook.

In both cases, I would like to understand the cost associated with sponsoring this development and the estimated time required for its implementation.

Could you please let me know the preferred communication channel for discussing these matters?

I appreciate your response in advance.

**Comment by remdex** [comment]
Hi,

You do not need to develop anything.  Please take a look at telegram or any other third party extension.
You can take a look what events I'm listening here. You can listen for the same events in webhook section.
https://github.com/LiveHelperChat/telegram/blob/master/bootstrap/bootstrap.php#L12

E.g
![image](https://github.com/LiveHelperChat/livehelperchat/assets/1146085/55b32148-d799-4cca-9ecd-85d883cdea2f)


You can contact if need more info via Discord or Skype...

Here list.
https://doc.livehelperchat.com/docs/hooks

Or you can just use github search just :)

https://github.com/search?q=repo%3ALiveHelperChat%2Flivehelperchat+%22erLhcoreClassChatEventDispatcher%3A%3AgetInstance%28%29-%3Edispatch%22&type=code

**Comment by GuilhermeDeLucca** [comment]
Redenx, I'm not sure, but I've been thinking about doing something relatively simple:

1. Using the API, create a chat:

`https://LHC/site_admin/restapi/chat
{
    "ignore_required": true,
    "ignore_bot": false,
    "department": [1],
    "fields": {
        "Username": "Visitor"
    },
    "messages": [
        {
            "msg": "Bot message",
            "time": 1691973690,
            "user_id": -2
        }
    ]
}
`

I successfully create a chat this way!

2. Create this chat as if it were through WebBook (simulating that it was done through WhatsApp). This way, I can respond to clients or start a conversation for a WhatsApp number!

`https://LHC/webhooks/incoming/open-wa
{
    "event": "onMessage",
    "data": {
        "id": "12345678911@c.us_KJDSUFSD&4564asdSADF",
        "viewed": false,
        "body": "Hi",
        "type": "chat",
        "from": "12345678911@c.us",
        "to": "12345678911@c.us",
        "author": "12345678911@c.us",
        "self": "in",
        "ack": 1,
        "invis": false,
        "isNewMsg": true,
        "star": false,
        "kicNotified": false,
        "recvFresh": true,
        "isFromTemplate": false,
        "pollInvalidated": false,
        "isSentCagPollCreation": false,
        "latestEditMsgKey": null,
        "latestEditSenderTimestampMs": null,
        "mentionedJidList": [],
        "groupMentions": [],
        "isVcardOverMmsDocument": false,
        "isForwarded": false,
        "labels": [],
        "hasReaction": false,
        "productHeaderImageRejected": false,
        "lastPlaybackProgress": 0,
        "isDynamicReplyButtonsMsg": false,
        "isCarouselCard": false,
        "parentMsgId": null,
        "isMdHistoryMsg": false,
        "stickerSentTs": 0,
        "isAvatar": false,
        "lastUpdateFromServerTs": 0,
        "invokedBotWid": null,
        "bizBotType": null,
        "botResponseTargetId": null,
        "botPluginType": null,
        "botPluginReferenceIndex": null,
        "botPluginSearchProvider": null,
        "botPluginSearchUrl": null,
        "botPluginMaybeParent": false,
        "botReelPluginThumbnailCdnUrl": null,
        "botMsgBodyType": null,
        "requiresDirectConnection": null,
        "fromMe": false,
        "isOnline": false,
        "participantsCount": 1
    },
    "isOnline": false,
    "chatId": "12345678911@c.us",
    "mediaData": [],
    "text": "Hi"
}
`

However, using this method, I can open a chat in the webhook or send a message, but I don't receive the chat ID in return, which means I can't manipulate it to add a subject or an operator to it."

![Captura de Tela 2024-01-30 às 12 10 44](https://github.com/LiveHelperChat/livehelperchat/assets/9898902/c85716f4-c0a9-4b0f-9850-0a46fc459597)

![Captura de Tela 2024-01-30 às 12 11 07](https://github.com/LiveHelperChat/livehelperchat/assets/9898902/a6e39f3e-f302-4102-ae4d-3fbe4781bf89)
Basically, what I want to do is:

Create a field for the user (non-administrator) where they will provide their phone number and send a message (thus initiating a message in the webhook, creating a chat). With this, I can obtain the ID of the opened chat and perform the interactions I need.

**Reply by remdex** [reply]
Chats created through webhook has much more info with them. PErhaps I'll extend webhook call to return chat id in that scenario.

**Comment by remdex** [comment]
I have pushed update it's enough now just append `?outout=json` to your webhook URL. In response you will get chat ID

More info here https://doc.livehelperchat.com/docs/development/incoming-webhooks#how-to-find-out-chat_id-if-im-sending-webhook-request

**Comment by GuilhermeDeLucca** [comment]
Hi Remdex,

Thank you for your message and the update! I appreciate the tip about finding the chat ID by adding "?output=json" to the webhook URL. I'll definitely check out the link you shared for more information.

By the way, with this update, do I need to update my system in any way to take advantage of this functionality? Or is it just a matter of making the change to the webhook URL?

I'm so pleased with this new feature that I'd be happy to buy you a couple of coffees (or more) as a token of my gratitude!

Thanks again for your help!

**Reply by remdex** [reply]
You can always make donation you want via github just :)

**Comment by AppDesv** [comment]
![Captura de Tela 2024-01-31 às 10 56 52](https://github.com/LiveHelperChat/livehelperchat/assets/142036461/4b6d9057-8ed5-4959-be17-b8656898d3c3)


here I still have no return

**Reply by remdex** [reply]
Update LHC and click clear cache in back office.
